File: edge-function/earnings_calendar.ts
/**
 * Supabase Edge Function: Earnings Calendar Fetcher
 * 
 * This Edge Function fetches earnings calendar data from StockTwits API,
 * transforms the data into a comprehensive earnings schedule format, and saves them to the database.
 * Accepts custom date ranges via POST request body.
 */

import { createClient, SupabaseClient } from 'https://esm.sh/@supabase/supabase-js@2.45.4'

// CORS headers for handling cross-origin requests
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'authorization, x-client-info, apikey, content-type',
  'Access-Control-Allow-Methods': 'POST, GET, OPTIONS, PUT, DELETE',
};

// Types for earnings calendar information
interface EarningsCalendar {
  symbol: string;
  exchange_id?: number;
  earnings_date: string;
  time_of_day?: string; // 'amc', 'bmo', 'dmh'
  eps?: number;
  eps_estimated?: number;
  eps_surprise?: number;
  eps_surprise_percent?: number;
  revenue?: number;
  revenue_estimated?: number;
  revenue_surprise?: number;
  revenue_surprise_percent?: number;
  fiscal_date_ending?: string;
  fiscal_year: number;
  fiscal_quarter?: number;
  market_cap_at_time?: number;
  sector?: string;
  industry?: string;
  conference_call_date?: string;
  conference_call_time?: string;
  webcast_url?: string;
  transcript_available?: boolean;
  status?: string;
  last_updated?: string;
  update_source?: string;
  data_provider: string;
  logo?: string;
}

// Request body interface
interface RequestBody {
  fromDate?: string; // Optional: YYYY-MM-DD format
  toDate?: string;   // Optional: YYYY-MM-DD format
  returnRawData?: boolean; // Optional: Return raw StockTwits format instead of saving to DB
}

// StockTwits API configuration for earnings calendar data
const STOCKTWITS_CONFIG = {
  name: 'StockTwits',
  baseUrl: 'https://api.stocktwits.com/api/2',
  endpoints: {
    earningsCalendar: '/discover/earnings_calendar',
  }
};

/**
 * Validate and parse date string
 */
function isValidDate(dateString: string): boolean {
  const regex = /^\d{4}-\d{2}-\d{2}$/;
  if (!regex.test(dateString)) return false;
  
  const date = new Date(dateString);
  return date instanceof Date && !isNaN(date.getTime());
}

/**
 * Get date range for earnings calendar
 * If dates are provided in request, use those. Otherwise, default to 1 month into the future.
 */
function getDateRange(requestBody?: RequestBody): { fromDate: string; toDate: string } {
  let fromDate: string;
  let toDate: string;
  
  if (requestBody?.fromDate && requestBody?.toDate) {
    // Validate provided dates
    if (!isValidDate(requestBody.fromDate) || !isValidDate(requestBody.toDate)) {
      throw new Error('Invalid date format. Use YYYY-MM-DD format.');
    }
    
    fromDate = requestBody.fromDate;
    toDate = requestBody.toDate;
    
    // Ensure fromDate is before toDate
    if (new Date(fromDate) > new Date(toDate)) {
      throw new Error('fromDate must be before toDate');
    }
  } else {
    // Default: today to 1 month in the future
    const today = new Date();
    const oneMonthLater = new Date(today);
    oneMonthLater.setMonth(today.getMonth() + 1);
    
    fromDate = today.toISOString().split('T')[0];
    toDate = oneMonthLater.toISOString().split('T')[0];
  }
  
  console.log(`Date range: ${fromDate} to ${toDate}`);
  
  return { fromDate, toDate };
}

/**
 * Fetch raw earnings calendar data from StockTwits API (preserves original structure)
 */
async function fetchRawStockTwitsData(fromDate: string, toDate: string): Promise<unknown> {
  try {
    const url = `${STOCKTWITS_CONFIG.baseUrl}${STOCKTWITS_CONFIG.endpoints.earningsCalendar}?date_from=${fromDate}&date_to=${toDate}`;
    console.log(`StockTwits: Fetching raw data from ${url}`);
    
    const response = await fetch(url);
    console.log(`StockTwits: Response status ${response.status}`);
    
    if (!response.ok) {
      console.log(`StockTwits: Response not ok: ${response.statusText}`);
      return null;
    }
    
    const data = await response.json();
    console.log('StockTwits: Raw response structure:', Object.keys(data));
    
    return data;
  } catch (error) {
    console.error(`StockTwits earnings calendar fetch error:`, error);
    return null;
  }
}

/**
 * Fetch earnings calendar data from StockTwits API and transform for database
 */
async function fetchFromStockTwits(fromDate: string, toDate: string): Promise<Partial<EarningsCalendar>[] | null> {
  try {
    const url = `${STOCKTWITS_CONFIG.baseUrl}${STOCKTWITS_CONFIG.endpoints.earningsCalendar}?date_from=${fromDate}&date_to=${toDate}`;
    console.log(`StockTwits: Fetching from ${url}`);
    
    const response = await fetch(url);
    console.log(`StockTwits: Response status ${response.status}`);
    
    if (!response.ok) {
      console.log(`StockTwits: Response not ok: ${response.statusText}`);
      return null;
    }
    
    const data = await response.json();
    console.log('StockTwits: Raw response structure:', Object.keys(data));
    
    // StockTwits API returns data nested by dates
    const allEarningsData: Array<{
      symbol: string;
      earnings_date: string;
      time: string;
      title: string;
      importance: number;
    }> = [];
    
    if (data.earnings && typeof data.earnings === 'object') {
      // Extract stocks from each date
      for (const [date, dateData] of Object.entries(data.earnings)) {
        const typedDateData = dateData as { stocks?: Array<{
          symbol: string;
          time: string;
          title: string;
          importance: number;
        }> };
        
        if (typedDateData?.stocks && Array.isArray(typedDateData.stocks)) {
          const dateStocks = typedDateData.stocks.map(stock => ({
            ...stock,
            earnings_date: date
          }));
          allEarningsData.push(...dateStocks);
        }
      }
    } else {
      console.log('StockTwits: Unexpected data structure - no earnings object found');
      return null;
    }
    
    console.log(`StockTwits: Found ${allEarningsData.length} total earnings records`);
    
    return allEarningsData.map(earning => ({
      symbol: earning.symbol,
      earnings_date: earning.earnings_date,
      time_of_day: parseTimeOfDay(earning.time),
      eps: undefined, // StockTwits doesn't provide EPS data in this endpoint
      eps_estimated: undefined,
      revenue: undefined, // StockTwits doesn't provide revenue data in this endpoint
      revenue_estimated: undefined,
      fiscal_year: new Date(earning.earnings_date).getFullYear(),
      fiscal_quarter: Math.ceil((new Date(earning.earnings_date).getMonth() + 1) / 3),
      status: 'scheduled',
      market_cap_at_time: undefined,
      sector: undefined,
      industry: undefined,
      conference_call_date: undefined,
      conference_call_time: earning.time,
      webcast_url: undefined,
      transcript_available: false,
      last_updated: new Date().toISOString(),
      update_source: 'stocktwits_api',
      data_provider: 'stocktwits'
    }));
    
  } catch (error) {
    console.error(`StockTwits earnings calendar fetch error:`, error);
    return null;
  }
}

/**
 * Parse time string to determine time of day
 */
function parseTimeOfDay(timeStr: string): string {
  if (!timeStr) return 'unknown';
  
  const hour = parseInt(timeStr.split(':')[0]);
  
  if (hour >= 16) return 'amc'; // After market close
  if (hour <= 9) return 'bmo';  // Before market open  
  return 'dmh'; // During market hours
}

/**
 * Combine earnings calendar data from multiple providers
 */
function combineEarningsData(dataArrays: (Partial<EarningsCalendar>[] | null)[]): EarningsCalendar[] {
  const validData = dataArrays.filter(data => data !== null) as Partial<EarningsCalendar>[][];
  
  if (validData.length === 0) return [];
  
  // Flatten all earnings records
  const allEarnings = validData.flat();
  
  // Group by symbol, fiscal_year, and fiscal_quarter to merge duplicates
  const earningsMap = new Map<string, EarningsCalendar>();
  
  for (const earning of allEarnings) {
    if (!earning.symbol || !earning.earnings_date || !earning.fiscal_year) continue;
    
    const key = `${earning.symbol}-${earning.fiscal_year}-${earning.fiscal_quarter || 'unknown'}`;
    
    if (earningsMap.has(key)) {
      // Merge with existing record, preferring non-null values
      const existing = earningsMap.get(key)!;
      const merged: EarningsCalendar = {
        symbol: earning.symbol,
        earnings_date: earning.earnings_date,
        fiscal_year: earning.fiscal_year,
        data_provider: `${existing.data_provider}, ${earning.data_provider}`
      };
      
      // Merge other fields, preferring non-null/non-empty values
      for (const [key, value] of Object.entries(earning)) {
        if (value !== null && value !== undefined && value !== '') {
          if (key !== 'data_provider' && !merged[key as keyof EarningsCalendar]) {
            (merged as EarningsCalendar & Record<string, unknown>)[key] = value;
          }
        }
      }
      
      // Calculate surprise percentages if we have both actual and estimated
      if (merged.eps && merged.eps_estimated && !merged.eps_surprise_percent) {
        merged.eps_surprise = merged.eps - merged.eps_estimated;
        merged.eps_surprise_percent = (merged.eps_surprise / merged.eps_estimated) * 100;
      }
      
      if (merged.revenue && merged.revenue_estimated && !merged.revenue_surprise_percent) {
        merged.revenue_surprise = merged.revenue - merged.revenue_estimated;
        merged.revenue_surprise_percent = (merged.revenue_surprise / merged.revenue_estimated) * 100;
      }
      
      earningsMap.set(key, merged);
    } else {
      // Create new record
      const newEarning: EarningsCalendar = {
        symbol: earning.symbol,
        exchange_id: earning.exchange_id,
        earnings_date: earning.earnings_date,
        time_of_day: earning.time_of_day,
        eps: earning.eps,
        eps_estimated: earning.eps_estimated,
        eps_surprise: earning.eps_surprise,
        eps_surprise_percent: earning.eps_surprise_percent,
        revenue: earning.revenue,
        revenue_estimated: earning.revenue_estimated,
        revenue_surprise: earning.revenue_surprise,
        revenue_surprise_percent: earning.revenue_surprise_percent,
        fiscal_date_ending: earning.fiscal_date_ending,
        fiscal_year: earning.fiscal_year,
        fiscal_quarter: earning.fiscal_quarter,
        market_cap_at_time: earning.market_cap_at_time,
        sector: earning.sector,
        industry: earning.industry,
        conference_call_date: earning.conference_call_date,
        conference_call_time: earning.conference_call_time,
        webcast_url: earning.webcast_url,
        transcript_available: earning.transcript_available || false,
        status: earning.status || 'scheduled',
        last_updated: new Date().toISOString(),
        update_source: 'multi_provider_fetch',
        data_provider: earning.data_provider || 'unknown'
      };
      
      // Calculate surprise percentages if we have both actual and estimated
      if (newEarning.eps && newEarning.eps_estimated && !newEarning.eps_surprise_percent) {
        newEarning.eps_surprise = newEarning.eps - newEarning.eps_estimated;
        newEarning.eps_surprise_percent = (newEarning.eps_surprise / newEarning.eps_estimated) * 100;
      }
      
      if (newEarning.revenue && newEarning.revenue_estimated && !newEarning.revenue_surprise_percent) {
        newEarning.revenue_surprise = newEarning.revenue - newEarning.revenue_estimated;
        newEarning.revenue_surprise_percent = (newEarning.revenue_surprise / newEarning.revenue_estimated) * 100;
      }
      
      earningsMap.set(key, newEarning);
    }
  }
  
  return Array.from(earningsMap.values());
}

/**
 * Save earnings calendar data to the database
 */
async function saveEarningsData(supabase: SupabaseClient, earningsData: EarningsCalendar[]): Promise<boolean> {
  if (earningsData.length === 0) return true;
  
  try {
    const { error } = await supabase
      .from('earnings_calendar')
      .upsert(earningsData, {
        onConflict: 'symbol,fiscal_year,fiscal_quarter,data_provider'
      });
    
    if (error) {
      console.error(`Error saving earnings calendar data:`, error);
      return false;
    }
    
    return true;
  } catch (error) {
    console.error(`Error in saveEarningsData:`, error);
    return false;
  }
}

/**
 * Main Edge Function handler
 */
Deno.serve(async (req) => {
  // Handle CORS preflight requests
  if (req.method === 'OPTIONS') {
    return new Response('ok', { headers: corsHeaders });
  }
  
  try {
    // Initialize Supabase client
    const supabaseClient = createClient(
      Deno.env.get('SUPABASE_URL') ?? '',
      Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? '',
    );
    
    console.log('Starting earnings calendar multi-provider fetch...');
    
    // Parse request body or query parameters for custom dates
    let requestBody: RequestBody | undefined;
    try {
      if (req.method === 'POST') {
        const contentType = req.headers.get('content-type');
        if (contentType?.includes('application/json')) {
          requestBody = await req.json();
        }
      } else if (req.method === 'GET') {
        // Parse query parameters for GET requests
        const url = new URL(req.url);
        const fromDateParam = url.searchParams.get('fromDate');
        const toDateParam = url.searchParams.get('toDate');
        const returnRawDataParam = url.searchParams.get('returnRawData');
        
        if (fromDateParam || toDateParam || returnRawDataParam) {
          requestBody = {
            fromDate: fromDateParam || undefined,
            toDate: toDateParam || undefined,
            returnRawData: returnRawDataParam === 'true'
          };
        }
      }
    } catch {
      console.log('No valid JSON body or query params provided, using default dates');
    }
    
    // Get date range (from request or use defaults)
    let fromDate: string;
    let toDate: string;
    
    try {
      const dateRange = getDateRange(requestBody);
      fromDate = dateRange.fromDate;
      toDate = dateRange.toDate;
    } catch (dateError) {
      return new Response(
        JSON.stringify({ 
          success: false, 
          error: (dateError as Error).message,
          message: 'Invalid date parameters'
        }),
        { 
          headers: { ...corsHeaders, 'Content-Type': 'application/json' },
          status: 400
        }
      );
    }
    
    console.log(`Fetching earnings calendar data from ${fromDate} to ${toDate}`);
    
    // Check if returnRawData is requested
    const returnRawData = requestBody?.returnRawData === true;
    
    try {
      console.log('Fetching earnings calendar data from StockTwits...');
      
      if (returnRawData) {
        // Fetch raw data and return in requested format
        const rawData = await fetchRawStockTwitsData(fromDate, toDate);
        
        if (rawData && typeof rawData === 'object' && 'earnings' in rawData) {
          const earningsData = rawData as { earnings?: Record<string, { stocks?: Array<{
            symbol: string;
            time: string;
            title: string;
            importance: number;
            emoji?: string;
          }> }> };
          
          // Transform to requested format
          const formattedEarnings: Record<string, { stocks: Array<{
            importance: number;
            symbol: string;
            date: string;
            time: string;
            title: string;
            emoji?: string;
          }> }> = {};
          
          if (earningsData.earnings) {
            for (const [date, dateData] of Object.entries(earningsData.earnings)) {
              if (dateData?.stocks && Array.isArray(dateData.stocks)) {
                formattedEarnings[date] = {
                  stocks: dateData.stocks.map(stock => ({
                    importance: stock.importance,
                    symbol: stock.symbol,
                    date: date,
                    time: stock.time,
                    title: stock.title,
                    ...(stock.emoji && { emoji: stock.emoji })
                  }))
                };
              }
            }
          }
          
          const response = {
            date_from: fromDate,
            date_to: toDate,
            earnings: formattedEarnings
          };
          
          return new Response(
            JSON.stringify(response),
            { 
              headers: { ...corsHeaders, 'Content-Type': 'application/json' },
              status: 200
            }
          );
        } else {
          return new Response(
            JSON.stringify({ 
              success: false, 
              message: 'StockTwits API returned no earnings data for the specified date range'
            }),
            { 
              headers: { ...corsHeaders, 'Content-Type': 'application/json' },
              status: 404
            }
          );
        }
      } else {
        // Original flow: fetch, transform, and save to database
        const stockTwitsData = await fetchFromStockTwits(fromDate, toDate);
        
        if (stockTwitsData && stockTwitsData.length > 0) {
          const combinedData = combineEarningsData([stockTwitsData]);
          
          if (combinedData.length > 0) {
            const saved = await saveEarningsData(supabaseClient, combinedData);
            
            if (saved) {
              // Group results by symbol for reporting
              const symbolCounts = new Map<string, number>();
              const processedSymbols = new Set<string>();
              
              combinedData.forEach(earning => {
                symbolCounts.set(earning.symbol, (symbolCounts.get(earning.symbol) || 0) + 1);
                processedSymbols.add(earning.symbol);
              });
              
              const results = Array.from(processedSymbols).map(symbol => ({
                symbol,
                status: 'success' as const,
                earnings_records: symbolCounts.get(symbol) || 0
              }));
              
              const response = {
                success: true,
                message: 'Earnings calendar fetch completed successfully',
                date_range: {
                  from: fromDate,
                  to: toDate
                },
                summary: {
                  total_symbols_found: processedSymbols.size,
                  total_earnings_records: combinedData.length
                },
                results: results.slice(0, 100) // Return up to 100 symbols
              };
              
              return new Response(
                JSON.stringify(response),
                { 
                  headers: { ...corsHeaders, 'Content-Type': 'application/json' },
                  status: 200
                }
              );
            } else {
              return new Response(
                JSON.stringify({ 
                  success: false, 
                  message: 'Failed to save earnings data to database'
                }),
                { 
                  headers: { ...corsHeaders, 'Content-Type': 'application/json' },
                  status: 500
                }
              );
            }
          } else {
            return new Response(
              JSON.stringify({ 
                success: false, 
                message: 'No valid earnings data found from StockTwits for the specified date range'
              }),
              { 
                headers: { ...corsHeaders, 'Content-Type': 'application/json' },
                status: 404
              }
            );
          }
        } else {
          return new Response(
            JSON.stringify({ 
              success: false, 
              message: 'StockTwits API returned no earnings data for the specified date range'
            }),
            { 
              headers: { ...corsHeaders, 'Content-Type': 'application/json' },
              status: 404
            }
          );
        }
      }
    } catch (error) {
      console.error('StockTwits earnings calendar fetch error:', error);
      return new Response(
        JSON.stringify({ 
          success: false, 
          message: `StockTwits API error: ${(error as Error).message}`
        }),
        { 
          headers: { ...corsHeaders, 'Content-Type': 'application/json' },
          status: 500
        }
      );
    }
    
  } catch (error) {
    console.error('Edge function error:', error);
    
    return new Response(
      JSON.stringify({ 
        success: false, 
        error: (error as Error).message,
        message: 'Internal server error'
      }),
      { 
        headers: { ...corsHeaders, 'Content-Type': 'application/json' },
        status: 500
      }
    );
  }
});

File: backend/src/service/market_engine/earnings_calendar.rs
use anyhow::{anyhow, Result};
use chrono::{Datelike, NaiveDate, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env;

use super::client::MarketClient;
use super::quotes::{get_quotes, get_simple_quotes};

/// Earnings calendar entry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EarningsCalendar {
    pub symbol: String,
    pub exchange_id: Option<u32>,
    pub earnings_date: String,
    pub time_of_day: Option<String>, // 'amc', 'bmo', 'dmh', 'unknown'
    pub eps: Option<f64>,
    pub eps_estimated: Option<f64>,
    pub eps_surprise: Option<f64>,
    pub eps_surprise_percent: Option<f64>,
    pub revenue: Option<f64>,
    pub revenue_estimated: Option<f64>,
    pub revenue_surprise: Option<f64>,
    pub revenue_surprise_percent: Option<f64>,
    pub fiscal_date_ending: Option<String>,
    pub fiscal_year: i32,
    pub fiscal_quarter: Option<u8>,
    pub market_cap_at_time: Option<f64>,
    pub sector: Option<String>,
    pub industry: Option<String>,
    pub conference_call_date: Option<String>,
    pub conference_call_time: Option<String>,
    pub webcast_url: Option<String>,
    pub transcript_available: Option<bool>,
    pub status: Option<String>,
    pub last_updated: Option<String>,
    pub update_source: Option<String>,
    pub data_provider: String,
    pub logo: Option<String>,
}

/// Request parameters for earnings calendar
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EarningsCalendarParams {
    pub from_date: Option<String>, // YYYY-MM-DD format
    pub to_date: Option<String>,   // YYYY-MM-DD format
    pub symbols: Option<Vec<String>>, // Optional filter by symbols
}

/// Edge Function response structure
#[derive(Debug, Deserialize)]
struct EdgeFunctionResponse {
    #[allow(dead_code)]
    date_from: String,
    #[allow(dead_code)]
    date_to: String,
    earnings: HashMap<String, EdgeFunctionDateData>,
}

#[derive(Debug, Deserialize)]
struct EdgeFunctionDateData {
    stocks: Vec<EdgeFunctionStock>,
}

#[derive(Debug, Deserialize)]
struct EdgeFunctionStock {
    #[allow(dead_code)]
    importance: u32,
    symbol: String,
    #[allow(dead_code)]
    date: String,
    time: String,
    #[allow(dead_code)]
    title: String,
    #[serde(default)]
    #[allow(dead_code)]
    emoji: Option<String>,
}

/// Request body for edge function
#[derive(Debug, Serialize)]
struct EdgeFunctionRequest {
    #[serde(rename = "fromDate")]
    from_date: String,
    #[serde(rename = "toDate")]
    to_date: String,
    #[serde(rename = "returnRawData")]
    return_raw_data: bool,
}

/// Parse time string to determine time of day
fn parse_time_of_day(time_str: &str) -> String {
    if time_str.is_empty() {
        return "unknown".to_string();
    }

    // Try to parse hour from time string (format: "HH:MM" or "HH:MM:SS")
    if let Some(hour_str) = time_str.split(':').next()
        && let Ok(hour) = hour_str.parse::<u32>() {
        if hour >= 16 {
            return "amc".to_string(); // After market close
        }
        if hour <= 9 {
            return "bmo".to_string(); // Before market open
        }
        return "dmh".to_string(); // During market hours
    }

    "unknown".to_string()
}

/// Parse market cap string to numeric value in millions
/// Handles formats like "200M", "1.5B", "500K", "$1.2B", etc.
fn parse_market_cap(market_cap_str: &str) -> Option<f64> {
    if market_cap_str.is_empty() {
        return None;
    }

    let cleaned = market_cap_str.trim().to_uppercase();
    
    // Remove common prefixes like "$"
    let cleaned = cleaned.trim_start_matches('$').trim();

    // Find the last character to determine the multiplier
    let (multiplier, numeric_part) = if let Some(stripped) = cleaned.strip_suffix('B') {
        (1000.0, stripped) // Billions to millions
    } else if let Some(stripped) = cleaned.strip_suffix('M') {
        (1.0, stripped) // Already in millions
    } else if let Some(stripped) = cleaned.strip_suffix('K') {
        (0.001, stripped) // Thousands to millions
    } else if let Some(stripped) = cleaned.strip_suffix('T') {
        (1_000_000.0, stripped) // Trillions to millions
    } else {
        // No suffix, assume millions
        (1.0, cleaned)
    };

    // Parse the number
    if let Ok(value) = numeric_part.trim().parse::<f64>() {
        Some(value * multiplier)
    } else {
        None
    }
}

/// Fetch market cap data for symbols using the quotes endpoint
async fn fetch_market_caps(
    client: &MarketClient,
    symbols: &[String],
) -> Result<HashMap<String, f64>> {
    let mut market_cap_map = HashMap::new();

    if symbols.is_empty() {
        return Ok(market_cap_map);
    }

    // Process in batches of 50 to avoid overwhelming the API
    const BATCH_SIZE: usize = 50;
    for chunk in symbols.chunks(BATCH_SIZE) {
        match get_quotes(client, chunk).await {
            Ok(quotes) => {
                for quote in quotes {
                    if let Some(market_cap_str) = &quote.market_cap
                        && let Some(market_cap_millions) = parse_market_cap(market_cap_str) {
                        market_cap_map.insert(quote.symbol.to_uppercase(), market_cap_millions);
                    }
                }
            }
            Err(e) => {
                log::warn!("Error fetching market caps for batch: {}", e);
                // Continue with next batch
            }
        }

        // Small delay between batches
        tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
    }

    log::info!("Fetched {} market caps", market_cap_map.len());
    Ok(market_cap_map)
}

/// Fetch earnings calendar data from Supabase Edge Function
async fn fetch_from_edge_function(
    from_date: &str,
    to_date: &str,
    symbols_filter: Option<&[String]>,
) -> Result<Vec<EarningsCalendar>> {
    // Get Supabase configuration from environment
    let supabase_url = env::var("SUPABASE_URL")
        .map_err(|_| anyhow!("SUPABASE_URL environment variable not set"))?;
    let supabase_anon_key = env::var("SUPABASE_ANON_KEY")
        .map_err(|_| anyhow!("SUPABASE_ANON_KEY environment variable not set"))?;

    // Construct edge function URL
    let edge_function_url = format!("{}/functions/v1/earnings-calendar", supabase_url);

    log::info!("Fetching earnings calendar from edge function: {}", edge_function_url);

    // Create HTTP client
    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(60))
        .build()?;

    // Prepare request body
    let request_body = EdgeFunctionRequest {
        from_date: from_date.to_string(),
        to_date: to_date.to_string(),
        return_raw_data: true,
    };

    // Make request to edge function
    let response = client
        .post(&edge_function_url)
        .header("Authorization", format!("Bearer {}", supabase_anon_key))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await?;

    let status = response.status();
    log::info!("Edge function response status: {}", status);

    if !status.is_success() {
        let body = response.text().await.unwrap_or_else(|_| "Unable to read body".to_string());
        log::error!("Error response body: {}", body);
        return Err(anyhow!(
            "Edge function returned status: {} - Body: {}",
            status,
            body
        ));
    }

    let body_text = response.text().await?;
    log::debug!("Response body length: {} bytes", body_text.len());

    let data: EdgeFunctionResponse = serde_json::from_str(&body_text)
        .map_err(|e| {
            log::error!("JSON parse error: {}", e);
            log::error!("Response body preview: {}", &body_text[..body_text.len().min(500)]);
            anyhow!("Failed to parse JSON: {} - Body preview: {}", e, &body_text[..body_text.len().min(200)])
        })?;

    let mut all_earnings: Vec<EarningsCalendar> = Vec::new();

    // Transform edge function response to EarningsCalendar format
    for (date, date_data) in data.earnings {
        for stock in date_data.stocks {
            // Filter by symbols if provided
            if let Some(symbols) = symbols_filter
                && !symbols.contains(&stock.symbol) {
                continue;
            }

            let earnings_date = NaiveDate::parse_from_str(&date, "%Y-%m-%d")
                .unwrap_or_else(|_| Utc::now().date_naive());
            let fiscal_year = earnings_date.year();
            let fiscal_quarter = Some(((earnings_date.month0() / 3) + 1) as u8);

            let earning = EarningsCalendar {
                symbol: stock.symbol,
                exchange_id: None,
                earnings_date: date.clone(),
                time_of_day: Some(parse_time_of_day(&stock.time)),
                eps: None,
                eps_estimated: None,
                eps_surprise: None,
                eps_surprise_percent: None,
                revenue: None,
                revenue_estimated: None,
                revenue_surprise: None,
                revenue_surprise_percent: None,
                fiscal_date_ending: None,
                fiscal_year,
                fiscal_quarter,
                market_cap_at_time: None,
                sector: None,
                industry: None,
                conference_call_date: None,
                conference_call_time: Some(stock.time),
                webcast_url: None,
                transcript_available: Some(false),
                status: Some("scheduled".to_string()),
                last_updated: Some(Utc::now().to_rfc3339()),
                update_source: Some("edge_function".to_string()),
                data_provider: "stocktwits".to_string(),
                logo: None,
            };

            all_earnings.push(earning);
        }
    }

    log::info!("Fetched {} earnings records from edge function", all_earnings.len());
    Ok(all_earnings)
}

/// Fetch logos for symbols using the simple quotes endpoint
async fn fetch_logos_for_symbols(
    client: &MarketClient,
    symbols: &[String],
) -> Result<HashMap<String, String>> {
    let mut logo_map = HashMap::new();

    if symbols.is_empty() {
        return Ok(logo_map);
    }

    // Process in batches of 50 to avoid overwhelming the API
    const BATCH_SIZE: usize = 50;
    for chunk in symbols.chunks(BATCH_SIZE) {
        match get_simple_quotes(client, chunk).await {
            Ok(quotes) => {
                for quote in quotes {
                    if let Some(logo) = quote.logo {
                        logo_map.insert(quote.symbol.to_uppercase(), logo);
                    }
                }
            }
            Err(e) => {
                log::warn!("Error fetching logos for batch: {}", e);
                // Continue with next batch
            }
        }

        // Small delay between batches
        tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
    }

    log::info!("Fetched {} logos", logo_map.len());
    Ok(logo_map)
}

/// Get earnings calendar data with optional date range and symbol filtering
/// Filters stocks to only include those with market cap above $200 million
pub async fn get_earnings_calendar(
    client: &MarketClient,
    params: EarningsCalendarParams,
) -> Result<Vec<EarningsCalendar>> {
    // Determine date range
    let (from_date, to_date) = if let (Some(from), Some(to)) = (params.from_date, params.to_date) {
        (from, to)
    } else {
        // Default to 1 month from today
        let today = Utc::now().date_naive();
        let to = today + chrono::Duration::days(30);
        (today.format("%Y-%m-%d").to_string(), to.format("%Y-%m-%d").to_string())
    };

    log::info!(
        "Fetching earnings calendar from {} to {}",
        from_date,
        to_date
    );

    // Fetch earnings data from edge function (without symbol filter first to get all symbols)
    let all_earnings = fetch_from_edge_function(
        &from_date,
        &to_date,
        params.symbols.as_deref(),
    )
    .await?;

    if all_earnings.is_empty() {
        return Ok(vec![]);
    }

    // Get unique symbols from earnings
    let unique_symbols: Vec<String> = all_earnings
        .iter()
        .map(|e| e.symbol.clone())
        .collect::<std::collections::HashSet<_>>()
        .into_iter()
        .collect();

    log::info!("Found {} unique symbols with earnings", unique_symbols.len());

    // Fetch market cap data for all symbols
    let market_cap_map = fetch_market_caps(client, &unique_symbols).await?;

    // Filter earnings to only include stocks with market cap > $200 million
    const MIN_MARKET_CAP_MILLIONS: f64 = 200.0;
    let mut filtered_earnings: Vec<EarningsCalendar> = all_earnings
        .into_iter()
        .filter(|earning| {
            let symbol_upper = earning.symbol.to_uppercase();
            if let Some(market_cap) = market_cap_map.get(&symbol_upper) {
                *market_cap > MIN_MARKET_CAP_MILLIONS
            } else {
                // If market cap is not available, exclude the stock (conservative approach)
                log::debug!("Excluding {} - market cap not available", earning.symbol);
                false
            }
        })
        .collect();

    log::info!(
        "Filtered to {} earnings records with market cap > ${}M",
        filtered_earnings.len(),
        MIN_MARKET_CAP_MILLIONS
    );

    // Update earnings with market cap data
    for earning in &mut filtered_earnings {
        let symbol_upper = earning.symbol.to_uppercase();
        if let Some(market_cap) = market_cap_map.get(&symbol_upper) {
            earning.market_cap_at_time = Some(*market_cap);
        }
    }

    // Fetch logos for filtered symbols
    if !filtered_earnings.is_empty() {
        let filtered_symbols: Vec<String> = filtered_earnings
            .iter()
            .map(|e| e.symbol.clone())
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect();

        let logo_map = fetch_logos_for_symbols(client, &filtered_symbols).await?;

        // Update earnings with logos
        for earning in &mut filtered_earnings {
            if let Some(logo) = logo_map.get(&earning.symbol.to_uppercase()) {
                earning.logo = Some(logo.clone());
            }
        }
    }

    Ok(filtered_earnings)
}